{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pandas == 0.24.2\n",
      "pyBigWig == 0.3.18\n",
      "pyfaidx == 0.7.0\n",
      "numpy == 1.16.6\n",
      "biopython == 1.76\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import pyBigWig\n",
    "from pyfaidx import Fasta\n",
    "import numpy as np\n",
    "from Bio.Seq import Seq\n",
    "import os\n",
    "print(\"pandas == {}\".format(pd.__version__))\n",
    "print(\"pyBigWig == {}\".format(pyBigWig.__version__))\n",
    "print(\"pyfaidx == {}\".format(__import__('pyfaidx').__version__))\n",
    "print(\"numpy == {}\".format(np.__version__))\n",
    "print(\"biopython == {}\".format(__import__('Bio').__version__))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 1. Extract shape signals from bw files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read annotation file\n",
    "df = pd.read_csv('./genomes/hg38.gff3', sep='\\t', comment='#', header=None,\n",
    "                 names=['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_info_pos = df[\n",
    "    (df['type'].isin(['processed_transcript', 'transcript','mRNA','miRNA','lincRNA','snRNA','RNA','snoRNA','rRNA'])) & \n",
    "    (df['strand']==\"+\")\n",
    "]\n",
    "transcript_id_pos = transcript_info_pos['attributes'].str.extract('ID=transcript:([^;]+);')[0].unique()\n",
    "len(transcript_id_pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "70317"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_info_neg = df[\n",
    "    (df['type'].isin(['processed_transcript', 'transcript','mRNA','miRNA','lincRNA','snRNA','RNA','snoRNA','rRNA'])) & \n",
    "    (df['strand']==\"-\")\n",
    "]\n",
    "transcript_id_neg = transcript_info_neg['attributes'].str.extract('ID=transcript:([^;]+);')[0].unique()\n",
    "len(transcript_id_neg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENST00000238081 is in transcript_info_neg.\n"
     ]
    }
   ],
   "source": [
    "# checking individual genes \n",
    "transcript_to_check = 'ENST00000238081'\n",
    "\n",
    "if transcript_to_check in transcript_id_neg:\n",
    "    print \"{} is in transcript_info_neg.\".format(transcript_to_check)\n",
    "else:\n",
    "    print \"{} is NOT in transcript_info_neg.\".format(transcript_to_check)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Positive strand, invivo\n",
    "genome = Fasta('./genomes/seq/hg38.fa')\n",
    "# Read the file that contain the SHAPE score\n",
    "bw_path = \"./shape_bw_human/icSHAPE_hg38_Cell_2016_NAI-N3_invivo_plus_score_vivo_NAIN3-score-pos.bw\"\n",
    "bw = pyBigWig.open(bw_path)\n",
    "\n",
    "for transcript_id in transcript_id_pos:\n",
    "    # Filter exon data for the current transcript_id\n",
    "    exon_data = df[(df['type'] == 'exon') & (df['attributes'].str.contains(str(transcript_id)))]\n",
    "    # Extract the SHAPE score from the positive strands, without separating the exons in a list\n",
    "    chrom_lengths = bw.chroms()\n",
    "    # Extract the exon coordinates\n",
    "    exon_coordinates = [(row.iloc[0], row.iloc[3], row.iloc[4]) for index, row in exon_data.iterrows()]\n",
    "    # List to store signal arrays for each exon\n",
    "    shape_sig = []\n",
    "    # Iterate through each set of coordinates\n",
    "    for chrom, start, end in exon_coordinates:\n",
    "        if chrom in chrom_lengths and start < chrom_lengths[chrom] and end <= chrom_lengths[chrom]:\n",
    "            # Retrieve the signal values for each position within the specified region\n",
    "            signals = bw.values(chrom, start-1, end)  # Adjust for zero-based indexing\n",
    "            # Extend the one-dimensional array with the new signals\n",
    "            if signals:  # Check if signals are not None or empty\n",
    "                shape_sig.extend(signals)\n",
    "    if np.all(np.isnan(shape_sig)):\n",
    "        continue  # Skip to the next iteration if all values in shape_sig are NaN\n",
    "    # Check the number of valid signal, must be >= 10\n",
    "    nan_count = np.sum(np.isnan(shape_sig))\n",
    "    non_nan_count = len(shape_sig) - nan_count\n",
    "    if non_nan_count < 10:\n",
    "        continue\n",
    "    # Check how many NAs in between the signal region\n",
    "    shape_sig_array = np.array(shape_sig)\n",
    "    if len(shape_sig_array) > 0:\n",
    "        first_valid_index = np.where(~np.isnan(shape_sig_array))[0][0]\n",
    "        last_valid_index = np.where(~np.isnan(shape_sig_array))[0][-1]\n",
    "        # Calculate the total number of values and the number of NaNs in the specified range\n",
    "        valid_range = shape_sig_array[first_valid_index:last_valid_index + 1]\n",
    "        total_values = len(valid_range)\n",
    "        num_nans = np.isnan(valid_range).sum()\n",
    "        # Check if NaNs are <= 5% of the total values in the range, change it to a higher value if needed\n",
    "        if num_nans > 0.05 * total_values:\n",
    "            continue  # Skip to the next iteration if NaNs exceed 5% of the range\n",
    "    exon_sequences = \"\"\n",
    "    # Extract sequences for each exon\n",
    "    for chrom, start, end in exon_coordinates:\n",
    "        # Retrieve the sequence and concatenate it\n",
    "        sequence = genome[chrom][start-1:end].seq  # pyfaidx uses 0-based start, hence start-1\n",
    "        exon_sequences += sequence\n",
    "    data = [(sig, 0, seq) for sig, seq in zip(shape_sig, exon_sequences)]\n",
    "    # Create the DataFrame\n",
    "    df2 = pd.DataFrame(data, columns=['shape_sig', 'uncert', 'exon_seq'])\n",
    "    df2['shape_sig'].fillna(-999, inplace=True) # replace the nan signal with -999 (not available)\n",
    "    df2 = df2.iloc[first_valid_index:last_valid_index + 1]\n",
    "    df2.reset_index(drop=True, inplace=True)\n",
    "    df2.index += 1\n",
    "    # Create a new folder\n",
    "    output_dir = './shape_extract/whole_human_pos_invivo'\n",
    "    # Create the directory if it doesn't exist\n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "    # Define the output file path\n",
    "    output_path = os.path.join(output_dir, '{}.map'.format(transcript_id))\n",
    "    # Save the DataFrame to a file, named according to the current transcript_id\n",
    "    df2.to_csv(output_path, sep='\\t', index=True, header=False)\n",
    "\n",
    "bw.close()\n",
    "genome.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Negative strand, invivo\n",
    "genome = Fasta('./genomes/seq/hg38.fa')\n",
    "bw_path = \"./shape_bw_human/icSHAPE_hg38_Cell_2016_NAI-N3_invivo_minus_score_vivo_NAIN3-score-neg.bw\"\n",
    "bw = pyBigWig.open(bw_path)\n",
    "\n",
    "for transcript_id in transcript_id_neg:\n",
    "    # Filter exon data for the current transcript_id\n",
    "    exon_data = df[(df['type'] == 'exon') & (df['attributes'].str.contains(str(transcript_id)))]\n",
    "    # Extract the SHAPE score from the positive strands, without separating the exons\n",
    "    chrom_lengths = bw.chroms()\n",
    "    # Extract the exon coordinates\n",
    "    exon_coordinates = [(row.iloc[0], row.iloc[3], row.iloc[4]) for index, row in exon_data.iterrows()]\n",
    "    # List to store signal arrays for each exon\n",
    "    shape_sig = []\n",
    "    # Iterate through each set of coordinates\n",
    "    for chrom, start, end in exon_coordinates:\n",
    "        if chrom in chrom_lengths and start < chrom_lengths[chrom] and end <= chrom_lengths[chrom]:\n",
    "            # Retrieve the signal values for each position within the specified region\n",
    "            signals = bw.values(chrom, start-1, end)  # Adjust for zero-based indexing\n",
    "            # Extend the one-dimensional array with the new signals\n",
    "            if signals:  # Check if signals are not None or empty\n",
    "                shape_sig.extend(signals)\n",
    "    if np.all(np.isnan(shape_sig)):\n",
    "        continue  # Skip to the next iteration if all values in shape_sig are NaN\n",
    "    # Check the number of valid signal, must be >= 10\n",
    "    nan_count = np.sum(np.isnan(shape_sig))\n",
    "    non_nan_count = len(shape_sig) - nan_count\n",
    "    if non_nan_count < 10:\n",
    "        continue\n",
    "    # Check if how many NAs are there in  the signal\n",
    "    shape_sig_array = np.array(shape_sig)\n",
    "    if len(shape_sig_array) > 0:\n",
    "        first_valid_index = np.where(~np.isnan(shape_sig_array))[0][0]\n",
    "        last_valid_index = np.where(~np.isnan(shape_sig_array))[0][-1]\n",
    "        # Calculate the total number of values and the number of NaNs in the specified range\n",
    "        valid_range = shape_sig_array[first_valid_index:last_valid_index + 1]\n",
    "        total_values = len(valid_range)\n",
    "        num_nans = np.isnan(valid_range).sum()\n",
    "        # Check if NaNs are <= 5% of the total values in the range, change it to a higher value if needed\n",
    "        if num_nans > 0.05 * total_values:\n",
    "            continue  # Skip to the next iteration if NaNs exceed 5% of the range\n",
    "    exon_sequences = \"\"\n",
    "    # Extract sequences for each exon\n",
    "    for chrom, start, end in exon_coordinates:\n",
    "        # Retrieve the sequence and concatenate it\n",
    "        sequence = genome[chrom][start-1:end].seq  # pyfaidx uses 0-based start, hence start-1\n",
    "        exon_sequences += sequence\n",
    "    # Reverse the signal\n",
    "    shape_sig_R = shape_sig[::-1]\n",
    "    # Reverse complement of the sequence\n",
    "    seq_temp = Seq(exon_sequences)\n",
    "    exon_seq_RV = str(seq_temp.reverse_complement())\n",
    "    data = [(sig, 0, seq) for sig, seq in zip(shape_sig_R, exon_seq_RV)]\n",
    "    # Create the DataFrame\n",
    "    df2 = pd.DataFrame(data, columns=['shape_sig', 'uncert', 'exon_seq'])\n",
    "    df2['shape_sig'].fillna(-999, inplace=True) # replace the nan signal with -999 (not available)\n",
    "    df2 = df2.iloc[len(df2)-last_valid_index-1:len(df2)-first_valid_index]\n",
    "    df2.reset_index(drop=True, inplace=True)\n",
    "    df2.index += 1\n",
    "    # Save the DataFrame to a file, named according to the current transcript_id\n",
    "    # Create a new folder\n",
    "    output_dir = './shape_extract/whole_human_neg_invivo'\n",
    "    # Create the directory if it doesn't exist\n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "    # Define the output file path\n",
    "    output_path = os.path.join(output_dir, '{}.map'.format(transcript_id))\n",
    "    df2.to_csv(output_path, sep='\\t', index=True, header=False)\n",
    "\n",
    "bw.close()\n",
    "genome.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Positive strand, invitro\n",
    "genome = Fasta('./genomes/seq/hg38.fa')\n",
    "# Read the file that contain the SHAPE score\n",
    "bw_path = \"./shape_bw_human/icSHAPE_hg38_Cell_2016_NAI-N3_invitro_plus_score_vitro_NAIN3-score-pos.bw\"\n",
    "bw = pyBigWig.open(bw_path)\n",
    "\n",
    "for transcript_id in transcript_id_pos:\n",
    "    # Filter exon data for the current transcript_id\n",
    "    exon_data = df[(df['type'] == 'exon') & (df['attributes'].str.contains(str(transcript_id)))]\n",
    "    # Extract the SHAPE score from the positive strands, without separating the exons in a list\n",
    "    chrom_lengths = bw.chroms()\n",
    "    # Extract the exon coordinates\n",
    "    exon_coordinates = [(row.iloc[0], row.iloc[3], row.iloc[4]) for index, row in exon_data.iterrows()]\n",
    "    # List to store signal arrays for each exon\n",
    "    shape_sig = []\n",
    "    # Iterate through each set of coordinates\n",
    "    for chrom, start, end in exon_coordinates:\n",
    "        if chrom in chrom_lengths and start < chrom_lengths[chrom] and end <= chrom_lengths[chrom]:\n",
    "            # Retrieve the signal values for each position within the specified region\n",
    "            signals = bw.values(chrom, start-1, end)  # Adjust for zero-based indexing\n",
    "            # Extend the one-dimensional array with the new signals\n",
    "            if signals:  # Check if signals are not None or empty\n",
    "                shape_sig.extend(signals)\n",
    "    if np.all(np.isnan(shape_sig)):\n",
    "        continue  # Skip to the next iteration if all values in shape_sig are NaN\n",
    "    nan_count = np.sum(np.isnan(shape_sig))\n",
    "    non_nan_count = len(shape_sig) - nan_count\n",
    "    if non_nan_count < 10:\n",
    "        continue\n",
    "    # Check how many NAs in between the signal region\n",
    "    shape_sig_array = np.array(shape_sig)\n",
    "    if len(shape_sig_array) > 0:\n",
    "        first_valid_index = np.where(~np.isnan(shape_sig_array))[0][0]\n",
    "        last_valid_index = np.where(~np.isnan(shape_sig_array))[0][-1]\n",
    "        # Calculate the total number of values and the number of NaNs in the specified range\n",
    "        valid_range = shape_sig_array[first_valid_index:last_valid_index + 1]\n",
    "        total_values = len(valid_range)\n",
    "        num_nans = np.isnan(valid_range).sum()\n",
    "        # Check if NaNs are <= 5% of the total values in the range, change it to a higher value if needed\n",
    "        if num_nans > 0.05 * total_values:\n",
    "            continue  # Skip to the next iteration if NaNs exceed 5% of the range\n",
    "    exon_sequences = \"\"\n",
    "    # Extract sequences for each exon\n",
    "    for chrom, start, end in exon_coordinates:\n",
    "        # Retrieve the sequence and concatenate it\n",
    "        sequence = genome[chrom][start-1:end].seq  # pyfaidx uses 0-based start, hence start-1\n",
    "        exon_sequences += sequence\n",
    "    data = [(sig, 0, seq) for sig, seq in zip(shape_sig, exon_sequences)]\n",
    "    # Create the DataFrame\n",
    "    df2 = pd.DataFrame(data, columns=['shape_sig', 'uncert', 'exon_seq'])\n",
    "    df2['shape_sig'].fillna(-999, inplace=True) # replace the nan signal with -999 (not available)\n",
    "    df2 = df2.iloc[first_valid_index:last_valid_index + 1]\n",
    "    df2.reset_index(drop=True, inplace=True)\n",
    "    df2.index += 1\n",
    "    # Create a new folder\n",
    "    output_dir = './shape_extract/whole_human_pos_invitro'\n",
    "    # Create the directory if it doesn't exist\n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "    # Define the output file path\n",
    "    output_path = os.path.join(output_dir, '{}.map'.format(transcript_id))\n",
    "    # Save the DataFrame to a file, named according to the current transcript_id\n",
    "    df2.to_csv(output_path, sep='\\t', index=True, header=False)\n",
    "\n",
    "bw.close()\n",
    "genome.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Negative strand, invitro\n",
    "genome = Fasta('./genomes/seq/hg38.fa')\n",
    "bw_path = \"./shape_bw_human/icSHAPE_hg38_Cell_2016_NAI-N3_invitro_minus_score_vitro_NAIN3-score-neg.bw\"\n",
    "bw = pyBigWig.open(bw_path)\n",
    "\n",
    "for transcript_id in transcript_id_neg:\n",
    "    # Filter exon data for the current transcript_id\n",
    "    exon_data = df[(df['type'] == 'exon') & (df['attributes'].str.contains(str(transcript_id)))]\n",
    "    # Extract the SHAPE score from the positive strands, without separating the exons\n",
    "    chrom_lengths = bw.chroms()\n",
    "    # Extract the exon coordinates\n",
    "    exon_coordinates = [(row.iloc[0], row.iloc[3], row.iloc[4]) for index, row in exon_data.iterrows()]\n",
    "    # List to store signal arrays for each exon\n",
    "    shape_sig = []\n",
    "    # Iterate through each set of coordinates\n",
    "    for chrom, start, end in exon_coordinates:\n",
    "        if chrom in chrom_lengths and start < chrom_lengths[chrom] and end <= chrom_lengths[chrom]:\n",
    "            # Retrieve the signal values for each position within the specified region\n",
    "            signals = bw.values(chrom, start-1, end)  # Adjust for zero-based indexing\n",
    "            # Extend the one-dimensional array with the new signals\n",
    "            if signals:  # Check if signals are not None or empty\n",
    "                shape_sig.extend(signals)\n",
    "    if np.all(np.isnan(shape_sig)):\n",
    "        continue  # Skip to the next iteration if all values in shape_sig are NaN\n",
    "    # Check the number of valid signal, must be >= 10\n",
    "    nan_count = np.sum(np.isnan(shape_sig))\n",
    "    non_nan_count = len(shape_sig) - nan_count\n",
    "    if non_nan_count < 10:\n",
    "        continue\n",
    "    # Check if how many NAs are there in  the signal\n",
    "    shape_sig_array = np.array(shape_sig)\n",
    "    if len(shape_sig_array) > 0:\n",
    "        first_valid_index = np.where(~np.isnan(shape_sig_array))[0][0]\n",
    "        last_valid_index = np.where(~np.isnan(shape_sig_array))[0][-1]\n",
    "        # Calculate the total number of values and the number of NaNs in the specified range\n",
    "        valid_range = shape_sig_array[first_valid_index:last_valid_index + 1]\n",
    "        total_values = len(valid_range)\n",
    "        num_nans = np.isnan(valid_range).sum()\n",
    "        # Check if NaNs are <= 5% of the total values in the range, change it to a higher value if needed\n",
    "        if num_nans > 0.05 * total_values:\n",
    "            continue  # Skip to the next iteration if NaNs exceed 5% of the range\n",
    "    exon_sequences = \"\"\n",
    "    # Extract sequences for each exon\n",
    "    for chrom, start, end in exon_coordinates:\n",
    "        # Retrieve the sequence and concatenate it\n",
    "        sequence = genome[chrom][start-1:end].seq  # pyfaidx uses 0-based start, hence start-1\n",
    "        exon_sequences += sequence\n",
    "    # Reverse the signal\n",
    "    shape_sig_R = shape_sig[::-1]\n",
    "    # Reverse complement of the sequence\n",
    "    seq_temp = Seq(exon_sequences)\n",
    "    exon_seq_RV = str(seq_temp.reverse_complement())\n",
    "    data = [(sig, 0, seq) for sig, seq in zip(shape_sig_R, exon_seq_RV)]\n",
    "    # Create the DataFrame\n",
    "    df2 = pd.DataFrame(data, columns=['shape_sig', 'uncert', 'exon_seq'])\n",
    "    df2['shape_sig'].fillna(-999, inplace=True) # replace the nan signal with -999 (not available)\n",
    "    df2 = df2.iloc[len(df2)-last_valid_index-1:len(df2)-first_valid_index]\n",
    "    df2.reset_index(drop=True, inplace=True)\n",
    "    df2.index += 1\n",
    "    # Save the DataFrame to a file, named according to the current transcript_id\n",
    "    # Create a new folder\n",
    "    output_dir = './shape_extract/whole_human_neg_invitro'\n",
    "    # Create the directory if it doesn't exist\n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "    # Define the output file path\n",
    "    output_path = os.path.join(output_dir, '{}.map'.format(transcript_id))\n",
    "    df2.to_csv(output_path, sep='\\t', index=True, header=False)\n",
    "\n",
    "bw.close()\n",
    "genome.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
